tasks/mesos.yml

```diff
@@ -19,62 +19,101 @@
-  notify:
-    - Reload daemon
-    - Restart mesos-slave
+  template: src=mesos-slave.service.j2 dest=/etc/systemd/system/mesos-slave.service
+  when: '{{     "slave" in mesos_install_mode
+            and systemd_check.stat.exists }}'
+  register: systemdunitslave
+
+# While waiting for https://github.com/ansible/ansible-modules-core/issues/191
+- name: Reload systemd
+  command: systemctl --system daemon-reload
+  when: '{{ ansible_service_mgr == "systemd" and (systemdunitmaster.changed|default(False) or systemdunitslave.changed|default(False)) }}'
 
 - name: Enable/Disable mesos-master service
-  service: name=mesos-master enabled="{{ (mesos_install_mode == 'slave') | ternary('no', 'yes') }}"
-  when: systemd_check.stat.exists == true
+  service:
+    name: mesos-master
+    enabled: yes
+  when: '{{ "master" in mesos_install_mode }}'
 
 - name: Enable/Disable mesos-slave service
-  service: name=mesos-slave enabled="{{ (mesos_install_mode == 'master') | ternary('no', 'yes') }}"
-  when: systemd_check.stat.exists == true
+  service:
+    name: mesos-slave
+    enabled: yes
+  when: '{{ "slave" in mesos_install_mode }}'
+
+- set_fact:
+    mesos_master_config_updated: '{{    default_config.changed|default(False)
+                                     or master_config.changed|default(False)
+                                     or (upstartdropinmaster.changed|default(False) and ansible_service_mgr == "upstart")
+                                     or (systemdunitmaster.changed|default(False)   and ansible_service_mgr == "systemd") }}'
+    mesos_slave_config_updated:  '{{    default_config.changed|default(False)
+                                     or slave_config.changed|default(False)
+                                     or (upstartdropinslave.changed|default(False)  and ansible_service_mgr == "upstart")
+                                     or (systemdunitslave.changed|default(False)    and ansible_service_mgr == "systemd") }}'
+    meta_cleanup:                '{{    mesos_slave_meta_cleanup|bool or mesos_agent_meta_cleanup|default(False)|bool }}'
+
+- debug:
+    msg: |
+      Forcing restart: {{ force_restart|bool }}.
+      Meta cleanup: {{ meta_cleanup|bool }}.
+
+- set_fact:
+    master_new_state: "{% if mesos_master_config_updated or force_restart|bool %}re{% endif %}started"
+  when: '{{ "master" in mesos_install_mode }}'
+  changed_when: True
+  notify:
+  - Control status of mesos-master service
+
+- debug: msg='Cleaning up meta/slaves/latest'
+  when: '{{     "slave" in mesos_install_mode
+            and (mesos_slave_config_updated or force_restart|bool)
+            and meta_cleanup|bool }}'
+  changed_when: True
+  notify:
+  - Cleanup meta/slaves/latest
+
+- set_fact:
+    slave_new_state: "{% if mesos_slave_config_updated or force_restart|bool %}re{% endif %}started"
+  when: '{{ "slave" in mesos_install_mode }}'
+  changed_when: True
+  notify:
+  - Control status of mesos-slave service
```

Comment:
👍  for re-using handlers 
though I was expecting to also get rid of this heavy  set_fact bool-expressions, in favor of using the different handlers (at least for the restarts) like before .. but maybe with the added logic, your solution is the better trade-off ?!